cogs/downloader.py

```diff
@@ -150,27 +149,73 @@ def save_repos(self):
-                               if cog != 'url' and
-                               self.repos[repo][cog]['INSTALLED'] is True]
-        for cog in installed_user_cogs:
-            await self.install(*cog)
-        await self.bot.say("Cogs updated. Reload all installed cogs? (yes/no)")
+
+        tasks = [self._update_repo(r) for r in self.repos]
+        tasknum = 0
+        num_repos = len(self.repos)
+
+        base_msg = "Downloading updated cogs, please wait... "
+        status = ' %d/%d repos updated' % (tasknum, num_repos)
+        msg = await self.bot.say(base_msg + status)
+
+        updated_cogs = []
+        new_cogs = []
+        deleted_cogs = []
+        installed_updated_cogs = []
+
+        for f in asyncio.as_completed(tasks, loop=self.bot.loop):
+            tasknum += 1
+            name, updates = await f
+            if updates:
+                for k, l in updates.items():
+                    if k == 'A':
+                        new_cogs.extend([(name, c) for c in l])
+                    elif k == 'D':
+                        deleted_cogs.extend([(name, c) for c in l])
+                    elif k == 'M':
+                        updated_cogs.extend([(name, c) for c in l])
+
+            status = ' %d/%d repos updated' % (tasknum, num_repos)
+            await self.bot.edit_message(msg, base_msg + status)
+        status = 'done. '
+
+        if not any(self.repos[repo][cog]['INSTALLED'] for repo, cog in updated_cogs):
+            status += ' No updates to apply. '
+
+        if new_cogs:
+            status += '\nNew cogs: ' + \
+                ', '.join('%s/%s' % c for c in new_cogs) + '.'
+        if deleted_cogs:
+            status += '\nDeleted cogs: ' + \
+                ', '.join('%s/%s' % c for c in deleted_cogs) + '.'
+        if updated_cogs:
+            status += '\nUpdated cogs: ' + \
+                ', '.join('%s/%s' % c for c in updated_cogs) + '.'
+
+        await self.bot.edit_message(msg, base_msg + status)
+
+        for repo, cog in updated_cogs:
+            if self.repos[repo][cog]['INSTALLED']:
+                installed_updated_cogs.append((repo, cog))
+                await self.install(repo, cog)
+
+        if not installed_updated_cogs:
+            return
+
+        await self.bot.say("Cogs updated. Reload updated installed cogs? (yes/no)")
         answer = await self.bot.wait_for_message(timeout=15,
                                                  author=ctx.message.author)
         if answer is None:
             await self.bot.say("Ok then, you can reload cogs with"
                                " `{}reload <cog_name>`".format(ctx.prefix))
         elif answer.content.lower().strip() == "yes":
-            for (repo, cog) in installed_user_cogs:
+            update_list = []
+            for (repo, cog) in installed_updated_cogs:
+                update_list.append(cog)
                 self.bot.unload_extension("cogs." + cog)
```

Comment:
I'd definitely surround lines 215-216 in a try block (and possibly think about calling `reload` from Owner instead of doing this directly)